design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,374 @@
+between BMH and Node objects. In other words, for the given set of label
+prefixes, the BMH is the source of truth and CAPM3 will enforce their
+presence on the Node.
+
+Additionally, there are certain well-known prefixes (e.g.
+kubelet.kubernetes.io/*, beta.kubernetes.io/*, kubernetes.io/*, k8s.io/*) that
+should be disallowed. Note the distinction between *.kubernetes.io/* (allowed)
+versus kubernetes.io/* (disallowed).
+
+### Synchronization Approach
+
+Synchronization must handle two scenarios: (i) a label is added/removed on a
+BMH and must be added/removed on the corresponding Node and (ii) a label is
+added/removed on the Node and must be removed/re-added on the Node to bring it
+in sync with the labels on BMH.
+
+The actual implementation of the synchronization logic may be encapsulated
+into an additional controller running within CAPM3; however, in the discussion
+below we simply reference CAPM3 as the entity doing the work.
+
+Scenario (i) can be handled as follows:
+
+1. CAPM3 will set up a watch on the BMH resource. It only requires read-only
+  access similar to what is done today:
+
+    baremetalhosts,verbs=get;list;watch
+
+2. On each BMH reconcile event, the controller will fetch the corresponding
+  Node. The _mapping_ can be accomplished by:
+
+    BMH.ConsumerRef --> Metal3Machine.OwnerRef --> Machine.Status.NodeRef
+
+3. Synchronize the labels and update the Node in the workload cluster. CAPM3
+  already fetches the Node objects from the workload clusters today using a
+  remote client. We would leverage that same capability here.
+
+An alternative means to accomplish the mapping is to use the ProviderID:
+
+    BMH.ConsumerRef --> Metal3Machine.Spec.ProviderID --> Node.Spec.ProviderID
+
+Unless we can set up filters, the approach would require us to fetch all Nodes
+and then scan through them to find the Node matching the ProviderID. This does
+not seem scalable for large clusters.
+Field-selectors could potentially be employed and would require further
+investigation. However,there is no obvious advantage over the earlier approach
+for doing this mapping.
+
+Scenario (ii) can be handled with a full-sync of all BMH and Node objects at a
+preconfigured synchronization interval:
+
+1. At a configurable interval (e.g. 15 seconds), fetch complete list of BMHs.
+2. For each BMH, fetch corresponding Node using the mapping discussed above.
+3. Synchronize the labels and update the Node in the workload cluster.
+
+An alternative would be to set up a watch on the Node resource types in every
+workload cluster. The benefit being that we can quickly synchronize any label
+changes on the Node object itself. However, this introduces complications when
+the management cluster != workload cluster and additionally, when there are
+many workload clusters being managed by a single management cluster. How such
+a watch could be implemented within the controller-runtime model would require
+further investigation.
+
+### Configuration
```

Comment:
There are several alternative approaches listed in this section, but it isn't clear which approach is really being proposed for use. I think it can't be the command-line flag approach. The annotation approach also seems to have been ruled out, is that right? Could you move the rejected approaches to the alternatives sections at the bottom of the document so it is clear?

Reply:
+1, annotation on one of the CAPM3 objects that applies to at least a whole MachinePool seems like the right way to configure this to me.